contracts/crashpunks-v2-friedger.clar

```diff
@@ -0,0 +1,345 @@
+
+;; nftIndex -> price (in ustx)
+;; if nftIndex is not in map, it is not listed for sale
+(define-map nft-market uint {price: uint})
+
+;; whitelist address -> # they can mint
+(define-map mint-pass principal uint)
+
+;; SIP-09: get last token id
+(define-read-only (get-last-token-id)
+  (ok (- (var-get mint-counter) u1))
+)
+
+;; SIP-09: URI for metadata associated with the token
+(define-read-only (get-token-uri (id uint))
+    (ok (some (var-get base-uri)))
+)
+
+;; SIP-09: Gets the owner of the 'Specified token ID.
+(define-read-only (get-owner (nftIndex uint))
+  (ok (nft-get-owner? crashpunks-v2 nftIndex))
+)
+
+;; SIP-09: Transfer
+(define-public (transfer (nftIndex uint) (owner principal) (recipient principal))
+    (begin
+        (asserts! (unwrap! (is-approved nftIndex contract-caller) ERR-NOT-AUTHORIZED) ERR-NOT-AUTHORIZED)
+        (asserts! (is-none (map-get? nft-market nftIndex)) ERR-NFT-LISTED)
+        (nft-transfer? crashpunks-v2 nftIndex owner recipient)
+    )
+)
+
+;; operable
+(define-read-only (is-approved (nftIndex uint) (operator principal))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (is-owned-or-approved nftIndex operator owner))
+    )
+)
+
+;; operable
+(define-public (set-approved (nftIndex uint) (operator principal) (approved bool))
+    (let ((owner (unwrap! (nft-get-owner? crashpunks-v2 nftIndex) ERR-COULDNT-GET-NFT-OWNER)))
+        (ok (map-set approvals {owner: contract-caller, operator: operator, nft-index: nftIndex} approved))
+    )
+)
+
+(define-public (set-approved-all (operator principal) (approved bool))
+    (ok (map-set approvals-all {owner: contract-caller, operator: operator} approved))
+)
+
+;; public methods
+
+;; upgrade from v1 to v2
+;; Owner of crashpunks v1 calls this upgrade function
+;; 1. Transfers v1 NFT to this contract
+;; 2. This contract mints the v2 NFT with the same nftIndex for contract-caller
+;; 3. This contract burns the v1 NFT
+(define-public (upgrade-v1-to-v2 (nftIndex uint))
+    ;; TODO: MAKE SURE THESE CONTRACT CALLS WORK, MAKE SURE THE CONRACT ADDRESSES WORKS FOR MAINNET
+    (begin 
+        ;; assert contract caller owns the v1 NFT at this nftIndex
+        (asserts! (is-eq contract-caller (unwrap! (unwrap! (contract-call? .crashpunks-v1 get-owner nftIndex) ERR-NOT-V1-OWNER) ERR-NOT-V1-OWNER)) ERR-NOT-V1-OWNER)
+
+        ;; 1. transfer v1 NFT to this contract
+        (try! (contract-call? .crashpunks-v1 transfer nftIndex contract-caller (as-contract tx-sender)))
+        
+        ;; 2. Mint the v2 NFT with the same nftIndex for contract-caller
+        (try! (nft-mint? crashpunks-v2 nftIndex contract-caller))
+
+        ;; 3. Burn the v1 NFT
+        (try! (contract-call? .crashpunks-v1 burn nftIndex (as-contract tx-sender)))
+        (ok true)
+    )
+)
+
+(define-public (batch-upgrade-v1-to-v2 (entries (list 200 uint)))
+    (ok (fold upgrade-v1-to-v2-helper entries false))
```

Comment:
If one fails the others can still succeed. Is that what we want?

Reply:
updated to fail entire tx if any fail